cpp/src/barretenberg/honk/sumcheck/sumcheck.test.cpp

```diff
@@ -402,4 +408,304 @@ TEST(Sumcheck, ProverAndVerifierLonger)
+
+    auto sumcheck_verifier = Sumcheck<Flavor, VerifierTranscript<FF>>(prover.key->circuit_size, verifier_transcript);
+
+    std::optional verifier_output = sumcheck_verifier.execute_verifier(relation_parameters);
+
+    ASSERT_TRUE(verifier_output.has_value());
+}
+
+/**
+ * @brief Test the Ultra Sumcheck Prover and Verifier for a real circuit
+ *
+ */
+TEST(Sumcheck, RealCircuitUltra)
+{
+    using Flavor = honk::flavor::Ultra;
+    using FF = typename Flavor::FF;
+    using ProverPolynomials = typename Flavor::ProverPolynomials;
+
+    // Create a composer and a dummy circuit with a few gates
+    auto composer = UltraHonkComposer();
+    fr a = fr::one();
+
+    // Add some basic add gates
+    uint32_t a_idx = composer.add_variable(a);
+    fr b = fr::one();
+    fr c = a + b;
+    fr d = a + c;
+    uint32_t b_idx = composer.add_variable(b);
+    uint32_t c_idx = composer.add_variable(c);
+    uint32_t d_idx = composer.add_variable(d);
+    for (size_t i = 0; i < 16; i++) {
+        composer.create_add_gate({ a_idx, b_idx, c_idx, 1, 1, -1, 0 });
+        composer.create_add_gate({ d_idx, c_idx, a_idx, 1, -1, -1, 0 });
+    }
+
+    // Add a big add gate with use of next row to test q_arith = 2
+    fr e = a + b + c + d;
+    uint32_t e_idx = composer.add_variable(e);
+
+    uint32_t zero_idx = composer.get_zero_idx();
+    composer.create_big_add_gate({ a_idx, b_idx, c_idx, d_idx, -1, -1, -1, -1, 0 }, true); // use next row
+    composer.create_big_add_gate({ zero_idx, zero_idx, zero_idx, e_idx, 0, 0, 0, 0, 0 }, false);
+
+    // Add some lookup gates (related to pedersen hashing)
+    barretenberg::fr pedersen_input_value = fr::random_element();
+    const fr input_hi = uint256_t(pedersen_input_value).slice(126, 256);
+    const fr input_lo = uint256_t(pedersen_input_value).slice(0, 126);
+    const auto input_hi_index = composer.add_variable(input_hi);
+    const auto input_lo_index = composer.add_variable(input_lo);
+
+    const auto sequence_data_hi = plookup::get_lookup_accumulators(plookup::MultiTableId::PEDERSEN_LEFT_HI, input_hi);
+    const auto sequence_data_lo = plookup::get_lookup_accumulators(plookup::MultiTableId::PEDERSEN_LEFT_LO, input_lo);
+
+    composer.create_gates_from_plookup_accumulators(
+        plookup::MultiTableId::PEDERSEN_LEFT_HI, sequence_data_hi, input_hi_index);
+    composer.create_gates_from_plookup_accumulators(
+        plookup::MultiTableId::PEDERSEN_LEFT_LO, sequence_data_lo, input_lo_index);
+
+    // Add a sort gate (simply checks that consecutive inputs have a difference of < 4)
+    a_idx = composer.add_variable(FF(0));
+    b_idx = composer.add_variable(FF(1));
+    c_idx = composer.add_variable(FF(2));
+    d_idx = composer.add_variable(FF(3));
+    composer.create_sort_constraint({ a_idx, b_idx, c_idx, d_idx });
+
+    // Add an elliptic curve addition gate
+    grumpkin::g1::affine_element p1 = crypto::generators::get_generator_data({ 0, 0 }).generator;
```

Comment:
why are the group elements from grumpkin if the scalars are from bn254? is this testing non-native arithmetic under the hood?

Reply:
This elliptic curve addition gate handles the addition of two points where one is scaled by the endomorphism constant and/or negated. My understanding of when this is useful is pretty vague (this example was copied directly from the elliptic add gate test in the composer tests). It's not clear to me why endomorphism constant is an fq (bberg::fr) -- this is firmly in @codygunton's wheelhouse, maybe he can shed some light